tidy3d/plugins/smatrix/component_modelers/terminal.py

```diff
@@ -249,3 +241,30 @@ def _check_grid_size_at_ports(
         yee_grid = simulation.grid.yee
         for port in ports:
             port._check_grid_size(yee_grid)
+
+    @staticmethod
+    def compute_power_waves(
+        port: Union[LumpedPort, CoaxialLumpedPort], sim_data: SimulationData
+    ) -> tuple[FreqDataArray, FreqDataArray]:
+        """Helper to compute the incident and reflected power waves at a port for a given
+        simulation result. The computed power waves have not been normalized.
+        """
+        voltage = port.compute_voltage(sim_data)
+        current = port.compute_current(sim_data)
+
+        power_a = (voltage + port.impedance * current) / 2 / np.sqrt(np.real(port.impedance))
+        power_b = (voltage - port.impedance * current) / 2 / np.sqrt(np.real(port.impedance))
+        return power_a, power_b
+
+    @staticmethod
+    def compute_power_delivered_by_port(
+        port: Union[LumpedPort, CoaxialLumpedPort], sim_data: SimulationData
+    ) -> FreqDataArray:
+        """Helper to compute the total power delivered to the network by a port for a given
+        simulation result. Units of power are Watts.
+        """
+        power_a, power_b = TerminalComponentModeler.compute_power_waves(
+            sim_data=sim_data, port=port
+        )
+        # Power delivered is the incident power minus the reflected power
+        return 0.5 * (np.abs(power_a) ** 2 - np.abs(power_b) ** 2)
```

Comment:
just wondering, if these are power already, why is it taking absolute value squared?  or is this another notion of power?

Reply:
that is a fair point. The equation is correct, but technically `power_a` is the amplitude of a "power wave". I'll choose a better name. FYI, the amplitudes are proportional to voltage and current:

![image](https://github.com/flexcompute/tidy3d/assets/155486263/9ad11892-2def-48e1-80cf-217cd0059841)